### PR TITLE
Added right fixed columns support

### LINKED ISSUE #2266: fixed columns right align
Hi,

For bootstrap-table-fixed-columns can you allow an option to select to fix the columns on the right or the left.

Thanks

### LINKED ISSUE #3480: Fixed column and detail view issue
Fixed column extension does not seem to work with "detail view" feature.

I used "Fixed column" extension to freeze the first 2 column for example
When i add "Detail View" ( where I can click on the plux (+) sign to view extra detail), the detail view will get overlapped each other and shifted down, also the fixed column will cover the detail view area.

fiddle link
http://jsfiddle.net/e3nk137y/17730/

basically these table tag does not seem to function together:
       data-detail-view="true"
       data-fixed-columns="true"
       data-fixed-number="2"

Example: 
<code>
<table data-toggle="table" 
       data-url="/gh/get/response.json/wenzhixin/bootstrap-table/tree/master/docs/data/data1/" 
       data-height="200"
       data-detail-view="true"
       data-fixed-columns="true"
       data-fixed-number="2">
    <thead>
    <tr>
        <th data-field="name">Name</th>
        <th data-field="stargazers_count">Stars</th>
        <th data-field="forks_count">Forks</th>
        <th data-field="description" class="description">Description</th>
    </tr>
    </thead>
</table>
</code>
I would be nice if they work together :)  ! . thanks,

### LINKED ISSUE #4510: Horizontal scrollbar obscured when using fixed columns
When using the fixed column plugin, the horizontal scrollbar is obscured (on Mac):

scrollbar-obscured

The height of ".fixed-table-body-columns" needs to be reduced by 15px to accommodate horizontal scrollbar.

### LINKED ISSUE #4534: fixed-columns has visual artifacts when there's enough room for columns anyway
When there exists enough horizontal space to display every column, fixed-columns isn't needed at all and should somehow "go away". Currently it displays visual artifacts.

[Online Example  (ensure your browser window is wide enough)](https://live.bootstrap-table.com/code/rparnas/399).

My workaround is to detect whether a horizontal scrollbar is present using the method from [here](https://stackoverflow.com/a/36900407/10048379):
````
var isScrollable = function(node) {
  var overflowY = window.getComputedStyle(node)['overflow-y'];
  var overflowX = window.getComputedStyle(node)['overflow-x'];
  return {
    vertical: (overflowY === 'scroll' || overflowY === 'auto') && node.scrollHeight > node.clientHeight,
    horizontal: (overflowX === 'scroll' || overflowX === 'auto') && node.scrollWidth > node.clientWidth,
  };
};
````

and to modify `fitHeader` to include:
````
. . .
this.$fixedBody.css({
  visibility: isScrollable(this.$tableBody[0]).horizontal ? "visible" : "hidden",
. . .
````

This may be related to #4510.

### LINKED ISSUE #4584: fixed columns select all problem
image

click the selectAll checkBox of fixed header, all rows has been selected, but the checkBox doesn't change

### LINKED ISSUE #4587: fixed-columns -> row mismatch on expand when scrollPosition <> 0
Using fixed-columns extension, if scrollPosition <> 0, the fixed columns mismatch the rest of the table. 

To reproduce the issue, scroll vertically through the table and click on a row.

The rows simply get back aligned when scrolling.

https://jsfiddle.net/alogg/fje895rv/26/

Additionally, an error is thrown by fixed-columns extension.
> TypeError: undefined is not an object (evaluating '$tr.attr('style').replace')

(An idea would be to refresh the table that contains the fixed columns on expand.)

Any help most appreciated, thanks in advance!
**GREAT TOOL GENTS !**

### LINKED ISSUE #4649: Fixed columns not working properly when there's nothing to scroll
Bootstrap-Table version: v1.15.5
Browsers: Chrome and Firefox

I'm using Bootstrap-Table with Bootstrap 4. My html composition is like this:
```
<div id="content-main">
   <div class="container" id="accordion">
            <div class="row">
                <div class="col-12"><table id="table" class="table order-list table-striped"></table>               
                </div>
            </div>
    </div>
</div>
```

In JS I initialize the table and I pull data from an API endpoint. I use a responseHandler to insert give that data the form that I need to show in the table. I have this:
```
$table = $('#table').bootstrapTable({
      method: 'get',
      url: '/api/endpoint/',
      responseHandler: 'responseHandler',
      classes: 'table table-hover',
      theadClasses: 'thead-dark header',
  });
```

When I process the new columns, I do this:
(Note that columns is a global variable in which I define my static structure of columns. Also, I set the responseHandler back because I had to call addColumns from my first responseHandler and that created an infinite loop).
```
function addColumns(values) {
  var new_column_header = {
    title: 'TEST',
    colspan: values.length
  };

  columns[0].push(new_column_header);
  for (value in values) {
      var new_column_value = {
        title: values[value][0],
        field: 'value_' + values[value][0],
        align: 'center',
      }
    }
    columns[1].push(new_column_value);
  }

  $table.bootstrapTable('refreshOptions', {
    columns: columns,
    fixedColumns: true,
    fixedNumber: 9,
    responseHandler: function (res) {
      return res;
    }
  });
}
```

Everything works fine. My table has 9 columns that are always there and I fix using the plug-in and then a few columns that I append dynamically. If the current browser size is wide enough so that all the table fits in, I get this weird interaction.
Pasted Graphic

If I make the browser window small enough that I can scroll at least a little bit, everything shows right.
Screenshot 2019-10-24 at 17 14 44

Please note that if I load the page with the window being maximized and then try to make it smaller, it will not work. It will stay like in the first picture.

### LINKED ISSUE #4775: fixed column plugin dislocation
image

hi, when I scale the browser to 80%,this fixed column layer does not in the correct position.how to fixed this issue?

### LINKED ISSUE #4797: FixedColumn + Column Checkbox does not work properly when use select all
The extension FixedColumn does not work properly using Column checkbox.

Only apply changes when resize the window, because click event not is set.

**Step 1**: Open [https://codepen.io/morojosa/pen/BayGdZV](https://codepen.io/morojosa/pen/BayGdZV)
**Step 2**: Select checkbox row 3 and row 4
**Step 3**: Resize window

### PR DESCRIPTION
**Pull request by wenzhixin**
* Fix #2266
* Fix #3480
* Fix #4510
* Fix #4534
* Fix #4584
* Fix #4587
* Fix #4649
* Fix #4775
* Fix #4797

Examples:

* Left and right without `height`: https://live.bootstrap-table.com/code/wenzhixin/1714
* Left and right with `height`: https://live.bootstrap-table.com/code/wenzhixin/1711
* Checkbox: https://live.bootstrap-table.com/code/wenzhixin/1713
* Auto enable fixed columns: https://live.bootstrap-table.com/code/wenzhixin/1712

### CODE REVIEW

**Comment by djhvscf** [comment]
Hi @wenzhixin  there are some css problems when height is set

File: src/extensions/fixed-columns/bootstrap-table-fixed-columns.js [hunk 1]
```diff
@@ -56,58 +51,244 @@ $.BootstrapTable = class extends $.BootstrapTable {
...
+
+  updateSelected () {
+    super.updateSelected()
+
+    if (!this.fixedColumnsSupported()) {
+      return
+    }
+
+    this.$tableBody.find('tr').each((i, el) => {
+      const index = $(el).data('index')
```
Comment (djhvscf) [inline src/extensions/fixed-columns/bootstrap-table-fixed-columns.js]: Can we save the $(el) in a const variable and use it in the next lines?

File: src/extensions/fixed-columns/bootstrap-table-fixed-columns.js [hunk 2]
```diff
@@ -56,58 +51,244 @@ $.BootstrapTable = class extends $.BootstrapTable {
...
 
+  initFixedColumnsHeader () {
+    if (this.options.height) {
+      this.needFixedColumns = this.$tableHeader.width() < this.$tableHeader.find('table').width()
+    } else {
+      this.needFixedColumns = this.$tableBody.width() < this.$tableBody.find('table').width()
+    }
+
+    if (this.needFixedColumns && this.options.fixedNumber) {
+      this.$fixedColumns.find('.fixed-table-header').remove()
```
Comment (djhvscf) [inline src/extensions/fixed-columns/bootstrap-table-fixed-columns.js]: Why removing .fixed-table-header and then find it again?

File: src/extensions/fixed-columns/bootstrap-table-fixed-columns.js [hunk 3]
```diff
@@ -56,58 +51,244 @@ $.BootstrapTable = class extends $.BootstrapTable {
...
+      this.$fixedColumns.append(this.$tableHeader.clone(true))
+      this.$fixedHeader = this.$fixedColumns.find('.fixed-table-header')
+
+      this.$fixedColumns.css({
+        width: this.getFixedColumnsWidth()
+      })
+    }
+
+    if (this.needFixedColumns && this.options.fixedRightNumber) {
+      this.$fixedColumnsRight.find('.fixed-table-header').remove()
```
Comment (djhvscf) [inline src/extensions/fixed-columns/bootstrap-table-fixed-columns.js]: Same here

File: src/extensions/fixed-columns/bootstrap-table-fixed-columns.js [hunk 4]
```diff
@@ -56,58 +51,244 @@ $.BootstrapTable = class extends $.BootstrapTable {
...
+        ) {
+          e.preventDefault()
+        }
+        this.$fixedBody.scrollTop(top)
+        this.$tableBody.scrollTop(top)
+        if (this.options.fixedRightNumber) {
+          this.$fixedBodyRight.scrollTop(top)
+        }
+      }).find('tr').hover(e => {
+        const index = $(e.currentTarget).data('index')
```
Comment (djhvscf) [inline src/extensions/fixed-columns/bootstrap-table-fixed-columns.js]: Can we save the e.currentTarget in a const variable and use it in next lines?

File: src/extensions/fixed-columns/bootstrap-table-fixed-columns.js (hunk 2 above)
Comment (UtechtDustin) [reply src/extensions/fixed-columns/bootstrap-table-fixed-columns.js]: The header will be recreated, thats the reason i guess.

Comment (UtechtDustin) [reply src/extensions/fixed-columns/bootstrap-table-fixed-columns.js]: This 3 lines will be used often, maybe create a function for that ?

**Comment by wenzhixin** [comment]
Finished this PR.

> there are some css problems when height is set

What problems?

**Comment by UtechtDustin** [comment]
Is ist compatible with all extensions ?

**Comment by wenzhixin** [comment]
@djhvscf windows + chrome?

**Comment by djhvscf** [comment]
Yes, sir

**Comment by UtechtDustin** [comment]
@wenzhixin 
Same for Linux + Firefox

**Comment by djhvscf** [comment]
If this helps

Windows + New Microsoft Edge

**Comment by UtechtDustin** [comment]
Please also check if this PR maybe fix #4510

**Comment by wenzhixin** [comment]
Finished, but the online example cannot work because of the cache.

**Comment by wenzhixin** [comment]
Fixed the bug of firefox and added all themes support:

* Bootstrap v3: https://live.bootstrap-table.com/code/wenzhixin/1715
* Bulma: https://live.bootstrap-table.com/code/wenzhixin/1716
* Semantic: https://live.bootstrap-table.com/code/wenzhixin/1717
* Materialize: https://live.bootstrap-table.com/code/wenzhixin/1718
* Foundation: https://live.bootstrap-table.com/code/wenzhixin/1719

**Comment by UtechtDustin** [comment]
Could you please update the examples int the first comment ? :-)

**Comment by UtechtDustin** [comment]
Looks like #4649 is also fixed with that PR, could you @wenzhixin please verify that ?

**Comment by wenzhixin** [comment]
> Could you please update the examples int the first comment ? :-)

updated

**Comment by djhvscf** [comment]
> > Could you please update the examples int the first comment ? :-)
> 
> updated

No more issues with css. Approved for my point of view

Thanks @wenzhixin

**Comment by wenzhixin** [comment]
> Looks like #4649 is also fixed with that PR, could you @wenzhixin please verify that ?

Yes.

**Comment by UtechtDustin** [comment]
I can also approve that PR, thank your @wenzhixin for the great PR :-)

**Pull request merged**